tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +227,280 @@ static bool HasCanonicalDimensionNumbers(
+// we can get the affine expression of dimensions that an
+// operand will access based on the input_str of einsum_config
+// For example:
+// let string_dim_umap = {'a' : d0, 'b' : d1, 'c' : d2}
+// for einsum_config "abc,cb->acb"
+// the first_input_operand will get umap[{"a","b","c"}] -> (d0, d1, d2)
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr, 4> getExprFromConfig(
+    std::set<std::string> input_set, std::vector<std::string> loop_dims,
+    std::unordered_map<std::string, AffineExpr> str_affine_dim_umap) {
+  SmallVector<AffineExpr, 4> OutputExpr;
+  for (std::string const& dim : loop_dims) {
+    OutputExpr.push_back(str_affine_dim_umap[dim]);
+  }
+  return OutputExpr;
+}
+
+// Given a real number n, will generate string of: "012...(n-1)"
+std::string rangeStr(size_t number) {
+  std::string range_str;
+  for (size_t i = 0; i < number; i++) range_str.append(std::to_string(i));
+  return range_str;
+}
+
+// Convert mhlo.einsum op into linalg.generic
+class EinsumToLinalgConverter : public OpConversionPattern<mhlo::EinsumOp> {
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    // Find maximum rank / number of loops.
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config().str();
+
+    // Setting Up Commonly Used Symbols as strings
+    std::string arrow = "->";
+    std::string comma = ",";
+    std::string ellipsis = "...";
+
+    // With the assumption of binary Input Operand and Single Output
+    // Get the inputs and output operands' indices
+    // einsum_config = "lhs_loop,rhs_loop->output_loop"
+    std::size_t pos_arrow = einsum_config.find(arrow);
+    std::size_t pos_comma = einsum_config.find(comma);
+
+    std::string lhs_loop = einsum_config.substr(0, pos_comma);
+    std::string rhs_loop = einsum_config.substr(
+        pos_comma + comma.size(), pos_arrow - (pos_comma + comma.size()));
+    std::string output_loop = einsum_config.substr(pos_arrow + arrow.size());
+
+    // Check for Invalid Configs
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 arrow
+    if (rhs_loop.find(comma) != std::string::npos ||
+        output_loop.find(comma) != std::string::npos ||
+        output_loop.find(arrow) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid Einsum Config!");
+    }
+
+    // Find result type, if on tensors.
+    Optional<ShapedType> result_ty;
+    result_ty = this->typeConverter->convertType(op->getResultTypes().front())
+                    .template dyn_cast<ShapedType>();
```

Comment:
Maybe `RankedShapedType` so we don't check `hasRank` below?

Reply:
I tried but, RankedShapedType seems to be iree specific? 🤔

Reply:
Sorry, I meant `RankedTensorType`

Reply:
You can specify the value directly and we don't need template.

```cpp
auto result_ty = ....
```